### PR TITLE
OG-59  Migrate to yarn workspaces and lerna

### PR DESCRIPTION
**Pull request by forshtat**
solves https://github.com/opengsn/gsn/issues/452 😀

### CODE REVIEW

File: packages/common/src/ContractInteractor.ts [hunk 1]
```diff
@@ -439,11 +425,11 @@ export default class ContractInteractor {
     return await this._getPastEvents(stakeManager.contract, names, extraTopics, options)
   }
 
-  // eslint-disable-next-line @typescript-eslint/require-await
   async _getPastEvents (contract: any, names: EventName[], extraTopics: string[], options: PastEventOptions): Promise<EventData[]> {
     const topics: string[][] = []
     const eventTopic = event2topic(contract, names)
     topics.push(eventTopic)
+    // TODO: AFAIK this means only the first parameter of the event is supported
```
[the remark this answers is not in the document]
Comment (forshtat) [reply packages/common/src/ContractInteractor.ts]: Not relevant to the PR

File: packages/common/src/ContractInteractor.ts [hunk 2]
```diff
@@ -605,6 +599,56 @@ export default class ContractInteractor {
   async hubDepositFor (paymaster: Address, transactionDetails: TransactionDetails): Promise<any> {
     return await this.relayHubInstance.depositFor(paymaster, transactionDetails)
   }
+
+  async resolveDeploymentVersions (): Promise<ObjectMap<PrefixedHexString>> {
```
[the remark this answers is not in the document]
Comment (forshtat) [reply packages/common/src/ContractInteractor.ts]: Not relevant to the PR

File: packages/common/src/EIP712/ForwarderUtil.ts [hunk 3]
```diff
@@ -19,10 +18,10 @@ export async function registerForwarderForGsn (forwarderTruffleOrWeb3: IForwarde
 
   function logTx (p: any): any {
     p.on('transactionHash', function (hash: string) {
-      log.debug(`Transaction broadcast: ${hash}`)
```
[the remark this answers is not in the document]
Comment (forshtat) [reply packages/common/src/EIP712/ForwarderUtil.ts:22]: Not relevant to the PR

File: packages/common/src/Environments.ts [hunk 4]
```diff
@@ -12,8 +12,8 @@ interface Environment {
 }
 
 export const defaultRelayHubConfiguration: RelayHubConfiguration = {
-  gasOverhead: 35965,
-  postOverhead: 13950,
+  gasOverhead: 35965 - 93,
+  postOverhead: 13637,
```
[the remark this answers is not in the document]
Comment (forshtat) [reply packages/common/src/Environments.ts:16]: Somehow even very minor changes in project structure make these values change. TBD.

File: packages/common/src/types/Aliases.ts [hunk 5]
```diff
@@ -44,3 +45,7 @@ export interface Web3ProviderBaseInterface {
     callback: (error: Error | null, result?: JsonRpcResponse) => void
   ) => void
 }
+
+export interface ObjectMap<T> {
```
[the remark this answers is not in the document]
Comment (forshtat) [reply packages/common/src/types/Aliases.ts]: Not relevant to the PR

File: packages/dev/contracts/Migrations.sol [hunk 6]
```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier:MIT
```
[the remark this answers is not in the document]
Comment (forshtat) [reply packages/dev/contracts/Migrations.sol:1]: In order to create truffle artefacts during testing. All contracts are in a different package now, so this is the only contract in the `dev` package.

File: packages/dev/test/provider/RelayClient.test.ts [hunk 7]
```diff
@@ -259,6 +259,8 @@ contract('RelayClient', function (accounts) {
         mockServer.use(bodyParser.urlencoded({ extended: false }))
         mockServer.use(bodyParser.json())
 
+        // used to work before workspaces, needs research
```
[the remark this answers is not in the document]
Comment (forshtat) [reply packages/dev/test/provider/RelayClient.test.ts:262]: Like, it has become more correct - `get` does not accept a method that returns a promise according to its type declaration. Of course we don't need the method to be async, so this is just a typing/linting error in a test file - did not want to touch the code lines in this PR as much as possible

File: packages/provider/package.json [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  "scripts": {
+    "tsc": "tsc",
+    "lint": "eslint -f unix .",
+    "rm-dist": "rm -rf tsconfig.tsbuildinfo dist"
+  },
+  "dependencies": {
+    "@opengsn/common": "^2.1.0",
+    "@types/body-parser": "^1.19.0",
+    "@types/eth-sig-util": "^2.1.0",
+    "@types/express-serve-static-core": "^4.17.18",
```
[the remark this answers is not in the document]
Comment (forshtat) [reply packages/provider/package.json:14]: I honestly don't know. I removed all dependencies and added when it complained. Now the list of dependencies in each package is small enough so we can actually review these manually.

File: package.json [hunk 9]
```diff
@@ -1,4 +1,15 @@
 {
+  "private": true,
+  "workspaces": {
+    "packages": [
+      "packages/*"
+    ],
+    "nohoist": [
```
[the remark this answers is not in the document]
Comment (forshtat) [reply package.json:7]: It didn't behave well for some reason. Nohoist is actually not that expensive for a small tool - just install it twice. Will remove if everything works fine down the road.,

File: package.json [hunk 10]
```diff
@@ -33,113 +44,46 @@
...
-    "typechain": "1.0.5",
-    "typechain-target-trufflehotfix": "0.0.4-alpha",
     "typescript": "3.8.3",
-    "web3-provider-engine": "14.1.0",
-    "webpack": "4.37.0",
-    "webpack-cli": "3.3.6"
-  },
-  "main": "dist/index.js",
-  "types": "dist/index.d.ts"
+    "web3": "1.2.6",
```
[the remark this answers is not in the document]
Comment (forshtat) [reply package.json:79]: To be safe 😀 Will remove later.

**Review by drortirosh — APPROVED** [review]

**Pull request merged**